python/ray/data/_internal/execution/interfaces.py

```diff
@@ -0,0 +1,267 @@
+
+    # Max number of in flight tasks. This is a soft limit.
+    parallelism_limit: Optional[int] = None
+
+    # Example: set to 1GB and executor will try to limit object store
+    # memory usage to 1GB. This is a soft limit.
+    memory_limit_bytes: Optional[int] = None
+
+    # Set this to prefer running tasks on the same node as the output
+    # node (node driving the execution).
+    locality_with_output: bool = False
+
+    # Always preserve ordering of blocks, even if using operators that
+    # don't require it.
+    preserve_order: bool = True
+
+
+class PhysicalOperator:
+    """Abstract class for physical operators.
+
+    An operator transforms one or more input streams of RefBundles into a single
+    output stream of RefBundles.
+
+    Operators are stateful and non-serializable; they live on the driver side of the
+    Dataset execution only.
+
+    Here's a simple example of implementing a basic "Map" operator:
+
+        class Map(PhysicalOperator):
+            def __init__(self):
+                self.active_tasks = []
+
+            def add_input(self, refs):
+                self.active_tasks.append(map_task.remote(refs))
+
+            def has_next(self):
+                ready, _ = ray.wait(self.active_tasks, timeout=0)
+                return len(ready) > 0
+
+            def get_next(self):
+                ready, remaining = ray.wait(self.active_tasks, num_returns=1)
+                self.active_tasks = remaining
+                return ready[0]
+
+    Note that the above operator fully supports both bulk and streaming execution,
+    since `add_input` and `get_next` can be called in any order. In bulk execution,
+    all inputs would be added up-front, but in streaming execution the calls could
+    be interleaved.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["PhysicalOperator"]):
+        self._name = name
+        self._input_dependencies = input_dependencies
+        for x in input_dependencies:
+            assert isinstance(x, PhysicalOperator), x
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def input_dependencies(self) -> List["PhysicalOperator"]:
+        """List of operators that provide inputs for this operator."""
+        assert hasattr(
+            self, "_input_dependencies"
+        ), "PhysicalOperator.__init__() was not called."
+        return self._input_dependencies
+
+    def get_stats(self) -> StatsDict:
+        """Return recorded execution stats for use with DatasetStats."""
+        raise NotImplementedError
+
+    def get_metrics(self) -> Dict[str, int]:
+        """Returns dict of metrics reported from this operator.
+
+        These should be instant values that can be queried at any time, e.g.,
+        obj_store_mem_allocated, obj_store_mem_freed.
+        """
+        return {}
+
+    def __reduce__(self):
+        raise ValueError("PhysicalOperator is not serializable.")
+
+    def __str__(self):
```

Comment:
we shall improve the string representation later to something like a tree structure

```
MapOperator[...]
  MapOperator[...]
    ReadOperator[...]
```

But this can be a followup if we find it useful.

Reply:
Sounds good.